core-bundle/src/Controller/AbstractFragmentController.php

```diff
@@ -155,8 +202,50 @@ protected function getType(): string
         return Container::underscore($className);
     }
 
-    protected function render(string $view, array $parameters = [], Response $response = null): Response
+    /**
+     * Add Contao default data to the template context.
+     *
+     * @param array<string>        $classes
+     * @param array<string, mixed> $properties
+     */
+    protected function addDefaultDataToTemplate(FragmentTemplate $template, Model $model, string $section, array $classes, array $properties): void
     {
+        if ($this->isLegacyTemplate($template->getName())) {
+            // Legacy fragments
+            $this->addHeadlineToTemplate($template, $model->headline);
+            $this->addCssAttributesToTemplate($template, $template->getName(), $model->cssID, $classes);
+            $this->addPropertiesToTemplate($template, $properties);
+            $this->addSectionToTemplate($template, $section);
+
+            return;
+        }
+
+        $headlineData = StringUtil::deserialize($model->headline ?: '', true);
+        $attributesData = StringUtil::deserialize($model->cssID ?: '', true);
+
+        $template->setData([
+            'type' => $this->getType(),
+            'model' => $model->row(),
+            'section' => $section,
+            'properties' => $properties,
+            'attributes' => (new HtmlAttributes())
+                ->setIfExists('id', $attributesData[0] ?? null)
+                ->addClass($attributesData[1] ?? '', ...$classes),
+            'headline' => [
+                'value' => $headlineData['value'] ?? '',
+                'unit' => $headlineData['unit'] ?? 'h1',
+            ],
```

Comment:
`value` and `unit` is not very intuitive IMO, how about

```suggestion
            'headline' => $headlineData['value'] ?? '',
            'headlineTag' => $headlineData['unit'] ?? 'h1',
```

or maybe even

```suggestion
            'headline' => $headlineData['value'] ?? '',
            'headlineLevel' => (int) substr($headlineData['unit'] ?? 'h1', 1),
```

Reply:
IMHO the value for the headline needs to be an array, so that you can easily pass it around - and then "headline" is already in the names: `headline.value`, `headline.unit`. 

Use it like so…

```twig
{% include "<your-headline-partial>" with {headline} %}
```

… with an element like so:
```twig
<{{ headline.unit }}>{{ headline.value }}</{{ headline.unit }}>
```

Reply:
I was wondering if having an `HtmlTag` object would make sense. You already provided us with the attributes class so maybe that would make the collection complete? :)

Reply:
I actually had thought about a class for tags as well. But when implementing the templates, I noticed that most of the times you benefit from helper objects when:
- there are multiple parts that belong together but should be passed along together (e.g. the `Figure`)
- operations building parts of it are expensive and should be lazy (e.g. a "Download" containing previews that are built on the fly)
- there are a lot of parameters/variants where IDE autocompletion is a big plus

IMHO grouping things into objects only makes sense if the templates have the same level of abstraction: e.g. a `_figure.html.twig` renders a `Figure` object, a `_headline.html.twig` renders a `Headline` object. Then it''s clear what to pass to this template if you're using it standalone. How the markup itself is looking is completely up to the template. 

I feel classes for HTML tags most of the time break this abstraction because they are too specific - here is an example for two versions of a headline that were built with the same context but a different template:

```html
<h1 class="headline">foo</h1>

<div class="headline"><h1><span class="first">f</span>oo</h1></div>
```

I think I would leave this part as is for now but go over the contexts of all templates again in the upcoming PR for the fragments. Then we can probably see best where abstractions are needed. Wdyt?